exercises/binary-search-tree/example.h

```diff
@@ -0,0 +1,226 @@
+
+        explicit binary_tree(T data)
+            : _data(std::move(data)),
+              _left(nullptr),
+              _right(nullptr)
+        {}
+
+        ~binary_tree() = default;
+
+        binary_tree(binary_tree &) = delete;
+        binary_tree& operator=(binary_tree &) = delete;
+
+        binary_tree(binary_tree &&) = delete;
+        binary_tree&& operator=(binary_tree &&) = delete;
+
+    public:
+        void insert(const T &data);
+
+        const T &data() const {return _data;};
+        const binary_tree_ptr& left() const {return _left;};
+        const binary_tree_ptr& right() const {return _right;};
+
+        binary_tree_iter begin() const;
+        binary_tree_iter end() const;
+        
+    private:
+        void insert(const T &data, binary_tree_ptr &location);
+        T _data;
+        binary_tree_ptr _left;
+        binary_tree_ptr _right;
+
+    public:
+        class binary_tree_iter final
+        {
+            friend binary_tree_iter binary_tree::begin() const;
+            friend binary_tree_iter binary_tree::end() const;
+        public:
+            using binary_tree_iter_ptr = std::unique_ptr<binary_tree_iter>;
+        private:
+            enum class state
+            {
+                Left,
+                Right,
+                Middle,
+                Done
+            };
+
+            const binary_tree &_tree;
+            state _state;
+            binary_tree_iter_ptr _side_iter;
+
+            explicit binary_tree_iter(const binary_tree &tree);
+            static binary_tree_iter build_end_iterator(const binary_tree &tree);
+
+        public:
+            ~binary_tree_iter() = default;
+
+            binary_tree_iter(binary_tree_iter &) = delete;
+            binary_tree_iter & operator=(binary_tree_iter &) = delete;
+
+            binary_tree_iter(binary_tree_iter &&) = default;
+            binary_tree_iter & operator=(binary_tree_iter &&) = default;
+
+            const T& operator*() const;
+            const T* operator->() const;
+            binary_tree_iter& operator++();
+
+            bool operator==(const binary_tree_iter &) const;
+            bool operator!=(const binary_tree_iter &) const;
+        
+        private:
+            void advance_side_iter(state next_state);
+            static binary_tree_iter_ptr build_first_side_iter(const binary_tree &tree);
+            static binary_tree_iter_ptr copy_side_iter(const binary_tree_iter &other);
+        };
+    };
+
+    // tree
+
+    template<typename T>
+    void binary_tree<T>::insert(const T &data, binary_tree<T>::binary_tree_ptr &location)
+    {
+        if (!location)
+            location = binary_tree_ptr(new binary_tree(data));
+        else
+            location->insert(data);   
+    }
+
+    template<typename T>
+    void binary_tree<T>::insert(const T &data)
+    {
+        binary_tree_ptr &insert_location = data > _data ? _right : _left;
+        insert(data, insert_location);
```

Comment:
This is the only place you call the private insert function. It's not doing anything logically different than this function so you should just place it inline.